crates/core_simd/src/vector.rs

```diff
@@ -311,6 +312,57 @@ where
         unsafe { self.store(slice.as_mut_ptr().cast()) }
     }
 
+    #[must_use]
+    #[inline]
+    pub fn masked_load_or(slice: &[T], or: Self) -> Self
+    where
+        Mask<<T as SimdElement>::Mask, N>: ToBitMask + ToBitMaskArray,
+    {
+        Self::masked_load_select(slice, Mask::splat(true), or)
+    }
+
+    #[must_use]
+    #[inline]
+    pub fn masked_load_select(
+        slice: &[T],
+        mut enable: Mask<<T as SimdElement>::Mask, N>,
+        or: Self,
+    ) -> Self
+    where
+        Mask<<T as SimdElement>::Mask, N>: ToBitMask + ToBitMaskArray,
+    {
+        enable &= {
+            let mask = bzhi_u64(u64::MAX, core::cmp::min(N, slice.len()) as u32);
+            let mask_bytes: [u8; 8] = unsafe { core::mem::transmute(mask) };
+            let mut in_bounds_arr = Mask::splat(true).to_bitmask_array();
+            let len = in_bounds_arr.as_ref().len();
+            in_bounds_arr.as_mut().copy_from_slice(&mask_bytes[..len]);
+            Mask::from_bitmask_array(in_bounds_arr)
+        };
```

Comment:
The previous version was generating suboptimal code. This implementation is still a bit gnarly because I wasn't able to easily cast a `u64` into a `Mask<_, N>` easily. This is because `ToBitmask` uses u8/u16/u32/u64. I don't think we can pull in `num_traits` for an easy solution to this?

This version generates code that matches my handwritten AVX-512 intrinsics.

Reply:
I made some (not yet merged) changes to bitmasks in #375 that may help you (with that change, it's possible to always return a `u64`), but I'm suspicious about this implementation, since it's not particularly portable.  Bitmasks are really only efficient on x86, and most x86 for the time being will not be AVX-512.  I'm confident this codegen will be horrendous on aarch64 :)

I think you should use the old version of the bounds checking, and conditionally check for x86 and N < 64 to use bitmasks.  I would actually split off the x86-optimized enable-adjusting code to its own function.

Reply:
I'll check it out. So far, the bitmasks haven't been a problem for me because of other codegen issues.

I hoped that LLVM implements the fallback by breaking the vector in half, log2 on every step. I was imagining something like this:

```rust
fn load_select<T, const N: usize> (slice: &[T], enable: Mask<T::Mask, N>, or: Simd<T, N>) -> Simd<T, N> where ... {
    if likely(enable.all() && in_bounds(N-1)) {
      Self::from_slice_unchecked(slice)
    } else {
        let low = Simd::<T, N/2>::load_select(slice, enable.low_half());
        let high = Simd::<T, N/2>::load_select(&slice[N..], enable.high_half());
        merge(low, high)
    }
}
```

But instead it generates code that's more like this:

```rust
fn load_select<T, const N: usize> (slice: &[T], enable: Mask<T::Mask, N>, or: Simd<T, N>) -> Simd<T, N> where ... {
    let mut elems = [T::default(); N];
    for lane_ix in 0..N {
        if in_bounds(lane_ix) && enable.test(lane_ix) {
            elems[lane_ix] = slice[lane_ix];
        }
    }
    Self::from_array(elems)
}
```

If I wanted to implement this for other architectures efficiently, could I do it with some form of recursive swizzle & macros? I think the potential for writing simpler code while retaining efficiency is there. The intended use case is to have the `likely(enable.all() && in_bounds(N-1))` branch be always taken, except for the last iteration. I think that would lead to good codegen, since you don't need to duplicate loop bodies if you wanted an epilogue, and it should also be perfectly predictable until the last iteration?

This could still be inefficient in certain cases:
* Even for AVX-512, not needing to do bitmask operations in all but one iteration could free up the core. In my experience, you can probably spare scalar resources for this so it doesn't seem like a big deal
* The caller depends on the mask for their computation after `load_select`. This could be a problem if they compute the mask differently than this library, and the compiler doesn't notice. Basically computing the masks twice, in different ways

Reply:
I don't think const generics are powerful enough to allow anything recursive, yet.  I think you're being too tricky for the compiler--if you'd like to skip the checks, you need to add the explicit `if enable.all()` branch.

Reply:
This was obviously easier for me to implement, since I didn't write it generically, but does this look right?
https://rust.godbolt.org/z/TvYqPh4js

First, I couldn't convince the compiler to emit bzhi so I had to resort to asm.  But even then--the naive approach seems more efficient to me.

Reply:
Check out my comment on StackOverflow here: https://stackoverflow.com/questions/75179720/how-to-get-rust-compiler-to-emit-bzhi-instruction-without-resorting-to-platform#comment136619035_75179720

I found two ways to emit bzhi but neither are production-ready IMO.

I will benchmark this some more. I like your constexpr array solution. When I did something similar before (but using a `&'static [usize; 64]`), it loaded the entire slice into multiple vector registers before shuffling.

Your solution is precise and indeed looks clean here, including for `-C target-feature=+sse4.1,+avx2`  version. `mm_maskload` comes from AVX/AVX2. Not sure about SSE & NEON but it's promising.

I don't think it's valid to do `len as i32`, you'd need something like `i32::try_from(len).unwrap_or(i32::MAX)` to do a saturating cast. Otherwise passing in a slice of length `i32::MAX as usize + 1` could make the bounds check reject valid elements, right? This applies to smaller masks too, like i16 and i8.

As for const generics and `N/2`, I'll try adding just a single branch `if likely(enable.all() && in_bounds(N-1))` to see if there's a perf hit on AVX-512, because it's definitely going to be a big win for other targets.

Reply:
This is silly, but maybe you can rely on the optimizer here: https://rust.godbolt.org/z/TxPYhx6K4

You would could add that `index` function to one of the `Sealed` traits.